rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/CertificateTreeValidationService.java

```diff
@@ -255,189 +258,185 @@ private boolean isValidationRunCompleted(ValidationResult validationResult) {
 
-            validatedManifestEntries.v2().stream().map(childContext ->
-                    validateCertificateAuthority(trustAnchor, registeredRepositories, childContext, accumulator)
-            ).forEachOrdered(result::addAll);
+            CertificateAuthorityValidationResult validatedManifestEntries
+                    = validateManifestEntries(trustAnchor, registeredRepositories, context, manifestUri, manifest, crlUri, x509Crl);
+            result.addAll(validatedManifestEntries);
         } catch (Exception e) {
-            result.error(ErrorCodes.UNHANDLED_EXCEPTION, e.toString(), ExceptionUtils.getStackTrace(e));
+            validations.error(ErrorCodes.UNHANDLED_EXCEPTION, e.toString(), ExceptionUtils.getStackTrace(e));
         }
         return result;
     }
 
-    private Tuple2<ValidationResult, List<CertificateRepositoryObjectValidationContext>> validateManifestEntries(
-            TrustAnchor trustAnchor,
+    private CertificateAuthorityValidationResult validateManifestEntries(
+            TrustAnchorData trustAnchor,
+            final Map<URI, RpkiRepository> registeredRepositories,
             CertificateRepositoryObjectValidationContext context,
             URI manifestUri,
-            RpkiObject manifestObject,
             ManifestCms manifest,
             URI crlUri,
-            X509Crl x509Crl,
-            Accumulator accumulator
+            X509Crl x509Crl
     ) {
-        ValidationResult temporary = newValidationResult(context.getLocation());
-        List<ManifestEntryValidationResult> objectValidationResult = manifest.getFiles().entrySet().parallelStream()
-                .map(entry -> validateManifestEntry(trustAnchor, context, manifestUri, crlUri, x509Crl, entry))
-                .collect(toList());
-
-        objectValidationResult.stream().map(ManifestEntryValidationResult::getValidationResult).forEach(temporary::addAll);
-        if (validationConfig.isStrictValidation() && temporary.hasFailures()) {
-            return new Tuple2<>(temporary, Collections.emptyList());
+        CertificateAuthorityValidationResult result = manifest.getFiles().entrySet().parallelStream()
+                .map(entry -> validateManifestEntry(trustAnchor, registeredRepositories, context, manifestUri, crlUri, x509Crl, entry))
+                .collect(
+                        CertificateAuthorityValidationResult::empty,
+                        CertificateAuthorityValidationResult::addAll,
+                        CertificateAuthorityValidationResult::addAll
+                );
+
+        if (validationConfig.isStrictValidation() && result.getValidationResult().hasFailures()) {
+            // RFC 6486bis: drop all validated objects from this CA (and its children) since there
+            // was at least one validation failure on this manifest.
+            return CertificateAuthorityValidationResult.of(result.getValidationResult());
```

Comment:
Should we also include the validation results from the (dropped) child certificate authorities here? In this case we need to use `result.getAllValidationResults()`.

Reply:
I agree that it will create a lot of confusion to see errors about objects that are not really taken into account.